### PR TITLE
Mono.Dynamic.Interpreter: Avoid NRE when loop body is not a block.

### PR DESCRIPTION
**Pull request by dmunch**
Fixes https://bugzilla.xamarin.com/show_bug.cgi?id=45141

### CODE REVIEW

**Comment by akoeplinger** [comment]
Could you please add a test to https://github.com/mono/mono/tree/master/mcs/class/System.Core/Test/System.Linq.Expressions ?

**Comment by dmunch** [comment]
Sure! I'm just somewhole not able to run the test on the monotouch profile (which I need to in order to make the DLR kick in).

Executing `make run-test PROFILE=monotouch` on master or mono-4.6.0-branch gives me the following error message: 

```
CS     [monotouch] nunit.core.dll
LogCapture.cs(30,25): error CS0103: The name `ConfigurationSettings' does not exist in the current context
NUnitFramework.cs(410,7): error CS0103: The name `ConfigurationSettings' does not exist in the current context
TestRunnerThread.cs(86,5): error CS0103: The name `ConfigurationSettings' does not exist in the current context
```

Anything I'm ignoring?

**Comment by akoeplinger** [comment]
Hey, running tests for the monotouch profile isn't supported on desktop as it would mean you'd call into ios native APIs which wouldn't work.

You can try the mobile_static profile which uses the interpreter too. For that you need to run autogen.sh with the `--with-runtime_preset=mobile_static` and build again.

**Comment by akoeplinger** [comment]
You could also try this old commit I used for forcing the interpreter on net_4_x before we had a working mobile_static profile (since that one can sometimes be a bit finicky to use): https://github.com/akoeplinger/mono/commit/5189ad578bfb998e0aa06899a44cb419a005ca2f

**Comment by dmunch** [comment]
Hu this rabbit hole is getting deeper. ´moble_static´ boils out with a compilation error when I try to run the tests (forgot to copy the output, could reproduce if it would be of any help). Cherry-picking your commit on top of mono-4.6.0-branch yielded some conflicts (which looked easy to resolve), however compilation of mono then fails with loads of

```
 The type or namespace name 'Dynamic' does not exist in the namespace 'System'. Are you missing an assembly reference?
```

Ironically, there's also this Exception.

```
[ERROR] FATAL UNHANDLED EXCEPTION: Mono.CSharp.InternalErrorException: ../dlr/Runtime/Microsoft.Dynamic/Interpreter/Instructions/ControlFlowInstructions.cs(612,35): Microsoft.Scripting.Interpreter.CompiledLoopInstruction._compiledLoop ---> System.NullReferenceException: Object reference not set to an instance of an object
 at IKVM.Reflection.Signature.WriteType (IKVM.Reflection.Emit.ModuleBuilder module, IKVM.Reflection.Writer.ByteBuffer bb, IKVM.Reflection.Type type) <0x263f0a0 + 0x001aa> in <filename unknown>:0
```

So, basically I'm not able to run the tests I need. Isn't there an easier way to run the tests with the interpreter given that I'm on a Mac? Couldn't  I run the tests with the monotouch profile on the Simulator or something?

**Comment by akoeplinger** [comment]
> ´moble_static´ boils out with a compilation error when I try to run the tests (forgot to copy the output, could reproduce if it would be of any help).

Hmm, interesting, it should work as it runs on our CI. Can you share the error please?

> Couldn't I run the tests with the monotouch profile on the Simulator or something?

Yes-ish, that would require using https://github.com/xamarin/xamarin-macios (which is not that easy if you're not familiar with the repo).

If you want you can also just paste the test here and I'll check it on the interpreter 😄

**Comment by dmunch** [comment]
Ok I'm one step further, actually I built with `--with-runtime_preset=mobile_static` but didn't specify `--with-mobile_static=yes` so he couldn't find the framework assemblies.

However I'm now stuck here: 

```
jack@DESKTOP-KT8HO74:/Volumes/MacintoshHD/Users/jack/mono-fork/mcs/class/System.Core$ make run-test PROFILE=mobile_static
ok=:; /Applications/Xcode.app/Contents/Developer/usr/bin/make run-test-recursive || ok=false; /Applications/Xcode.app/Contents/Developer/usr/bin/make run-test-local || ok=false; $ok
cd ./../../tools/nunit-lite && /Applications/Xcode.app/Contents/Developer/usr/bin/make
/Applications/Xcode.app/Contents/Developer/usr/bin/make do-all TOP_LEVEL_DO=do-all
/Applications/Xcode.app/Contents/Developer/usr/bin/make all-local
make[5]: Nothing to be done for `all-local'.
/Applications/Xcode.app/Contents/Developer/usr/bin/make all-local
make[5]: Nothing to be done for `all-local'.
/Applications/Xcode.app/Contents/Developer/usr/bin/make all-local
make[4]: Nothing to be done for `all-local'.
/Applications/Xcode.app/Contents/Developer/usr/bin/make aot-all-profile
AOT     [mobile_static] AOT All Assemblies
make[3]: *** [aot-all-profile] Error 1
make[2]: *** [do-all-aot] Error 2
make[1]: *** [../../build/deps/nunit-mobile_static.stamp] Error 2
make: *** [do-run-test] Error 1
```

**Comment by marek-safar** [comment]
Fixed by using CoreFX code

**Comment by dmunch** [comment]
Great, thanks! Great to see that the whole Microsoft.Dynamic fork was dumped in favor of CoreFX.

**Comment by akoeplinger** [comment]
@Lapinou42 the "fix" (using the CoreFX code instead of the old DLR fork) isn't in the version you have.

@marek-safar do you think we should take this fix for 4.8 ?

**Comment by Lapinou42** [comment]
@akoeplinger Okay, I understand. I need to download the git of @dmunch and then build the Mono framework by myself for now to get this fix ? Right ?

I hope it will be available soon in Xamarin Studio because I have a lot of crashes in my app just for this. 

Thank you for your quick answer !

**Comment by dmunch** [comment]
@Lapinou42 I published a drop-in NuGet package which contains the fix. You can install it your iOS project until the official fix gets into a stable mono release. See details here: https://github.com/AutoMapper/AutoMapper/issues/1682#issuecomment-253742835

**Comment by Lapinou42** [comment]
@dmunch  Oh ! Great ! I test right now !

Thank you !

**Comment by Lapinou42** [comment]
Ok, it seems to works ! 
Unfortunatelly, I have now another crash in line 798 in the file /Library/Frameworks/Xamarin.iOS.framework/Versions/10.4.0.123/src/mono/mcs/class/dlr/Runtime/Microsoft.Scripting.Core/Ast/MethodCallExpression.cs : 

=> Parameter name "method" cannot be null.

Maybe another bug in the Mono framework.

Anyway, thank you for your nuget package ;)

**Comment by srinathganguri** [comment]
@akoeplinger @marek-safar Do you guys know what version of Xamarin stable release covers this fix? We are having problems with this exact same bug during our development environment due to which we have not upgraded our Xamarin.iOS libraries to 10 as the upgrade causes it and this has been for months. 

I work in the enterprise and we are worried to have the Mono framework build ourselves in the development environment to get by the issue as our build systems have to incorporate that change as well I assume. Please suggest as to which release of Xamarin this bug will be fixed?

**Comment by Lapinou42** [comment]
@srinathganguri It should be fixed in 15.2 release (It should be in preview at the beginning of next month)

See details of my bug report here on bugzilla: https://bugzilla.xamarin.com/show_bug.cgi?id=53435

So for now, just waiting the release like us.
